File: langchain_agent/repo-src/mastra_agent/src/agentFactory.ts
import { Agent } from '@mastra/core/agent';
import { createOpenAI } from '@ai-sdk/openai';
import { createAnthropic } from '@ai-sdk/anthropic';
import { buildBffTools, type ToolSchema, type RunCtx, type EmitFn } from './bffToolAdapter';

const DEFAULT_INSTRUCTIONS =
  'You are a helpful banking assistant. Use the available tools to answer the user\'s questions.';

export interface LlmProviderConfig {
  baseUrl: string;
  apiKey: string;
  anthropicApiKey?: string;
  model: string;
  provider?: string;
}

export function buildAgent(
  toolSchemas: ToolSchema[],
  runCtx: RunCtx,
  llm: LlmProviderConfig,
  instructions?: string,
  emitFn?: EmitFn,
): Agent {
  const tools = buildBffTools(toolSchemas, runCtx, emitFn);

  const toolMap: Record<string, (typeof tools)[number]> = {};
  for (const tool of tools) {
    toolMap[tool.id] = tool;
  }

  let languageModel: unknown;
  if (llm.provider === 'anthropic') {
    // Use the native Anthropic provider with the Anthropic API key — NOT the
    // OpenAI-compat/LM-Studio key, which would 401 against api.anthropic.com.
    const anthropicProvider = createAnthropic({ apiKey: llm.anthropicApiKey ?? '' });
    const effectiveModel = llm.model.startsWith('claude') ? llm.model : 'claude-sonnet-4-6';
    languageModel = anthropicProvider(effectiveModel);
  } else {
    // createOpenAI() with a baseURL points the OpenAI provider at any
    // OpenAI-compatible endpoint (LM Studio, Groq, Together, etc).
    const openaiProvider = createOpenAI({ baseURL: llm.baseUrl, apiKey: llm.apiKey });
    languageModel = openaiProvider(llm.model);
  }

  // languageModel returns an AI SDK LanguageModel that Mastra's Agent
  // accepts at runtime, but its TS surface is DynamicArgument<MastraModelConfig>
  // which doesn't include the AI SDK type directly — cast through unknown.
  return new Agent({
    id: 'banking-agent',
    name: 'Banking Agent',
    instructions: instructions ?? DEFAULT_INSTRUCTIONS,
    model: languageModel as unknown as never,
    tools: toolMap,
  });
}

File: pydantic_agent/src/agent_factory.py
from __future__ import annotations
import os
from typing import Callable, Coroutine, Optional
from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIModel
from pydantic_ai.providers.openai import OpenAIProvider
from .bff_tool_adapter import build_tool_functions
from .models import BffDeps

_ANTHROPIC_DEFAULT_MODEL = "claude-sonnet-4-6"


def build_agent(
    tool_schemas: list[dict],
    model_name: str,
    base_url: str,
    api_key: str,
    system_prompt: str | None = None,
    provider: str = "",
    emit_fn: Optional[Callable[[dict], Coroutine]] = None,
) -> Agent:
    if provider == "anthropic":
        from pydantic_ai.models.anthropic import AnthropicModel
        from pydantic_ai.providers.anthropic import AnthropicProvider
        anthropic_key = os.environ.get("ANTHROPIC_API_KEY", "")
        effective_model = model_name if model_name else _ANTHROPIC_DEFAULT_MODEL
        # pydantic-ai >=1.x AnthropicModel has no api_key kwarg; the key is passed
        # via the provider. Passing api_key= directly raises TypeError on every run.
        model = AnthropicModel(
            effective_model,
            provider=AnthropicProvider(api_key=anthropic_key or None),
        )
    else:
        # Constructing OpenAIModel with an explicit provider keeps pydantic_ai from
        # falling back to the env-driven default (which would 401 on LM Studio).
        model = OpenAIModel(
            model_name=model_name,
            provider=OpenAIProvider(base_url=base_url, api_key=api_key),
        )
    tools = build_tool_functions(tool_schemas, emit_fn=emit_fn)
    prompt = system_prompt or (
        "You are a helpful banking assistant. "
        "Use the available tools to answer user questions accurately."
    )
    return Agent(
        model,
        deps_type=BffDeps,
        tools=tools,
        system_prompt=prompt,
        defer_model_check=True,
    )
